### PR TITLE
[OBP] support assuming roles to read secrets

### PR DESCRIPTION
**Pull request by jackie-ob**

### CODE REVIEW

**Comment by saikonen** [comment]
Other than my one comment on being able to set per-source roles, this seems to work as expected with the aws-secrets-provider at least 👍

File: metaflow/plugins/secrets/secrets_decorator.py [hunk 1]
```diff
@@ -195,14 +209,25 @@ def task_pre_step(
...
         for secret_spec_str_or_dict in self.attributes["sources"]:
             if isinstance(secret_spec_str_or_dict, str):
                 secret_specs.append(
-                    SecretSpec.secret_spec_from_str(secret_spec_str_or_dict)
+                    SecretSpec.secret_spec_from_str(secret_spec_str_or_dict, role=role)
                 )
             elif isinstance(secret_spec_str_or_dict, dict):
                 secret_specs.append(
-                    SecretSpec.secret_spec_from_dict(secret_spec_str_or_dict)
+                    SecretSpec.secret_spec_from_dict(secret_spec_str_or_dict, role=role)
```
[the remark this answers is not in the document]
Comment (jackie-ob) [reply metaflow/plugins/secrets/secrets_decorator.py:230]: For max granular control, we ought to support configuring "role" at multiple overlaying levels:
* global (mf config)
* `@secrets(.... role=)`
* `@secrets(sources=[{ .... role=}])`
This is the existing pattern we have taken with other settings like `METAFLOW_AWS_SECRETS_MANAGER_DEFAULT_REGION`, or `DEFAULT_SECRETS_BACKEND_TYPE`).

In this case we could follow the same pattern (not yet fully implemented here in this PR).

Here are the 2 reasons why we may not add per source control over role right now:
* Adding per-source just means slightly more code and more testing.  Not really huge, but just something.
* Multiple sources with different roles is a plausible use case, albeit much less common.  Note I think using different roles for different sources within a single cloud is more likely than multi-cloud.  In any case, it can be a nice mechanism for gathering feedback when someone needs this feature and asks about it.

In conclusion, my opinion is that if you have the cycles right now to add the per-source functionality, please go for it.  But we should not overly delay the release of the basic feature as is.  We can always follow up later.

**Review by saikonen — APPROVED** [review]
